pkg/kube/krt/README.md

```diff
@@ -0,0 +1,157 @@
+
+For example, a simple `Singleton` example that keeps track of the number of `ConfigMap`s in the cluster:
+
+```go
+ConfigMapCount := krt.NewSingleton[int](func(ctx krt.HandlerContext) *int {
+    cms := krt.Fetch(ctx, ConfigMaps)
+    return ptr.Of(len(cms))
+})
+```
+
+The `Fetch` operation enables querying against other collections.
+If the result of the `Fetch` operation changes, the collection will automatically be recomputed; the framework handles the state and event detection.
+In the above example, the provided function will be called (at least) every time there is a change to a configmap.
+The `ConfigMapCount` collection will produce events only when the count changes.
+The framework will automatically suppress events when nothing has changed.
+
+### Picking a collection type
+
+There are a variety of collection types available.
+Picking these is about simplicity, usability, and performance.
+
+The `NewSingleton` form (`func() *O`), in theory, could be used universally.
+Consider a transformation from `Pod` to `SimplePod`:
+
+```go
+SimplePods := krt.NewSingleton[SimplePod](func(ctx krt.HandlerContext) *[]SimplePod {
+    res := []SimplePod{}
+    for _, pod := range krt.Fetch(ctx, Pod) {
+        res = append(res, SimplePod{Name: pod.Name})
+    }
+    return &res
+}) // Results in a Collection[[]SimplePod]
+```
+
+While this *works*, it is inefficient and complex to write.
+Consumers of SimplePod can only query the entire list at once.
+Anytime *any* `Pod` changes, *all* `SimplePod`s must be recomputed.
+
+A better approach would be to lift `Pod` into a primary dependency:
+
+```go
+SimplePods := krt.NewCollection[SimplePod](func(ctx krt.HandlerContext, pod *v1.Pod) *SimplePod {
+    return &SimplePod{Name: pod.Name})
+}) // Results in a Collection[SimplePod]
+```
+
+Not only is this simpler to write, its far more efficient.
+Consumers can more efficiently query for `SimplePod`s using label selectors, filters, etc.
+Additionally, if a single `Pod` changes we only recompute one `SimplePod`.
+
+Above we have a one-to-one mapping of input and output.
+We may have one-to-many mappings, though.
+In these cases, usually its best to use a `ManyCollection`.
+Like the above examples, its *possible* to express these as normal `Collection`s, but likely inefficient.
+
+Example computing a list of all container names across all pods:
+```go
+ContainerNames := krt.NewManyCollection[string](func(ctx krt.HandlerContext, pod *v1.Pod) (res []string) {
+    for _, c := range pod.Spec.Containers {
+      res = append(res, c.Name)
+    }
+    return res
+}) // Results in a Collection[string]
+```
+
+Example computing a list of service endpoints.
+```go
+Endpoints := krt.NewManyCollection[Endpoint](func(ctx krt.HandlerContext, svc *v1.Service) (res []Endpoint) {
+    for _, c := range krt.Fetch(ctx, Pods, krt.FilterLabel(svc.Spec.Selector)) {
```

Comment:
Is this a good example? Why would we need to select svc pods instead of fetching endpoints

Reply:
Maybe a bad example, this was showing a simple example of actually implementing the k8s core EndpointsController. will replace it